### PR TITLE
Portable: create units with default SELinux label

### PR DESCRIPTION
**Pull request by cgzones**
Alternative to #20767

### CODE REVIEW

**Comment by bluca** [comment]
This is not an alternative to #20767 as it does not solve the same problem. Also on all my instances as far as I can see everything is already been labelled with the defaults, could you please explain why is it necessary to change to use these APIs? Are there corner cases where the right thing doesn't happen for instance?

**Comment by doverride** [comment]
I like this implementation better but I will leave it to others to choose.
I think the trade-offs made by the other implementation are not worth it.

**Comment by keszybz** [comment]
@doverride can you please describe what the intent of the patch is: i.e. essentially answer the questions that @bluca asked above. I'm trying to wrap my head around this, and while I think I understand why `open_tmpfile_linkable_label()` exists, I don't quite understand why `symlink_label()` exists and how it is different from `symlink()`.

**Comment by doverride** [comment]
This patch and the patch proposed by @bluca are mutually exclusive. This patch ensures that units copied from the image by portabled are created with the context specified on the host.

The patch proposed by @bluca does the opposite: it ensured that units copied from the image by portabled are created with the context associated with the unit on the image instead.

I do not particularly like the latter approach because to some extent it relies on the labels of the image. My argument is that files located on the host file system should be labeled according to the policy on the host for consistency.

The patch proposed by @bluca is also not going to fully address running SELinux portables because this only deals with unit files copied to the host and other aspects might still not be consistent between host and image. It was argued by others that this is "conflating" different issues but I do not see what the point is of implementing an arguable nasty hack to ensure that the portable unit has the label on the host that the image creator prefers over what the owner of the host prefers when this does currently not address other aspects. I.e the unit label on the host may or may not be usable, but then still other aspects might be inconsistent causing thing not to work after all. In my view, even though some say its conflating, that is just relevant because eventually it is about running the portable service and not about the label of the unit on the host.

I am also arguing that even though using portables with SELinux is not completely painless and transparent (probably never wiil be),  that we made nice progress by allowing users to customize context specifications on the host at runtime without requiring `systemctl daemon-reexec` or `systemctl daemon--reload` and so it fairly easy to configure the host to address portables with custom labeling in many cases.

There are also some other relatively minor operational issues with copying labels from images when it comes to attaching portables with --runtime.

Eventually if I cannot associate host labels associated with unit files copied portables from portables, then I intent to disallow portabled to manually label the units by blocking access to `setfscreate` (setfilecon) and rely on automatic labeling as defined by the policy on the host instead.

So in resume:

1. i prefer this solution over the solution proposed by @bluca
2. the two solutions are mutually exclusive because one solution wants to associate host labels with the copied units and the other solution wants to copy over the labels associated with the unit on the image.
3.  I will not oppose the patch proposed by @bluca because if that patch gets adopted instead of this then I can make it so that effectively nothing changes for me as it can just block `setfscreate` access and fall back to automatic labeling on the host.

File: src/portable/portable.c [hunk 1]
```diff
@@ -1126,7 +1128,7 @@ static int attach_unit_file(
                 _cleanup_(unlink_and_freep) char *tmp = NULL;
                 _cleanup_close_ int fd = -1;
 
-                fd = open_tmpfile_linkable(path, O_WRONLY|O_CLOEXEC, &tmp);
+                fd = open_tmpfile_linkable_label(path, O_WRONLY|O_CLOEXEC, &tmp);
```
Comment (bluca) [inline src/portable/portable.c:1131]: We've gone in a different direcion, so please drop this change - the rest is good

**Comment by bluca** [comment]
needs a rebase

File: src/shared/tmpfile-util-label.c [hunk 2]
```diff
@@ -24,3 +24,21 @@ int fopen_temporary_label(
 
         return r;
 }
+
+int open_tmpfile_linkable_label(
+                const char *target,
+                int flags,
+                char **ret_path) {
```
Comment (poettering) [inline src/shared/tmpfile-util-label.c:31]: hmm, do we really want individual wrappers around all these helper?

I'd prefer if we went down this route instead:
https://github.com/systemd/systemd/pull/20767#discussion_r714732194

**Comment by poettering** [comment]
hmm, can you explain why we should set the label from the database here? I mean, if i get this right for the files we create here we can get the label by 3 different means:

1. copy them from the image (which @bluca added for the unit files)
2. let the kernel determine them according to the rules of the loaded policy and the directory we create these files in
3. look them up in the database (which this PR does)

Why is 3 better than 2? Why doesn't 2 suffice?

(I'd be fine to to make the copying of the labels like @bluca added some you can opt-out of, via an env var, i.e. a secondary configuration option, mybe $SYSTEMD_SELINUX_LOCAL_LABELS=1 disables copying of the labels)

**Comment by doverride** [comment]
Lennart Poettering ***@***.***> writes:

> hmm, can you explain why we should set the label from the database here? I mean, if i get this right for the files
> we create here we can get the label by 3 different means:
>
> 1 copy them from the image (which @bluca added for the unit files)
> 2 let the kernel determine them according to the rules of the loaded policy and the directory we create these files
>  in
> 3 look them up in the database (which this PR does)
>
> Why is 3 better than 2? Why doesn't 2 suffice?

Because of confined service administration.

If you do 2. (like we do currently) then you have to tell the kernel (in
detail what label to associate with the unit) and that only works if you
have something to go by and in this case the unit is copied to a tmp
file and then renamed. so you cannot just tell the kernel: when portable
creates a unit in /etc/systemd/system.attached with name foo.service
then label that with this specified context automatically. (first of all
it doesnt create it in /etc/systemd/system.attached, and second of all
it doesnt create the unit with name foo.service. But even if, that
would still be cumbersome to maintain and expensive. But eventually we
want to be able to say: joe can start foo.service but not bar.service
and labels matter there (do not want to rely on polkit either)

with 3 all it takes is some selinux awareness and a context
specification. portable will take care of the rest and it doesnt matter
what the file is called or where it is in the interim aslong as
portabled knows what its doing (and @cgzones took good care of that)

Copying contexts from foreign images is a bad idea and i am not going to
allow portabled to do that. It was said: well this stuff already happens
with network filesystems and networking. But there is one considerable
difference: the network is not your local filesystem and a network
filesystem is not your local filesystem either. When that unit is copied
to the host, its the host policy that should have the last say.

>
> (I'd be fine to to make the copying of the labels like @bluca added some you can opt-out of, via an env var, i.e. a
> secondary configuration option, mybe $SYSTEMD_SELINUX_LOCAL_LABELS=1 disables copying of the labels)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android. 
>

-- 
gpg --locate-keys ***@***.***
Key fingerprint = FCD2 3660 5D6B 9D27 7FC6  E0FF DA7E 521F 10F6 4098
https://sks-keyservers.net/pks/lookup?op=get&search=0xDA7E521F10F64098
Dominick Grift

**Comment by bluca** [comment]
> Copying contexts from foreign images is a bad idea and i am not going to allow portabled to do that. It was said: well this stuff already happens with network filesystems and networking. But there is one considerable difference: the network is not your local filesystem and a network filesystem is not your local filesystem either. When that unit is copied to the host, its the host policy that should have the last say.

No, it's not a bad idea, as it has been explained a million times already, so please stop with this absolutist statements. It doesn't fit your use case with untrusted images, which is perfectly fine (apart from the tiny detail that filesystems from untrusted images can exploit your kernel and all that, but that's life on Linux) and you are entitled to set up your images and your policies as you see fit. But you are not entitled to break our use case, which was not just validated but specifically requested by our SELinux experts, just because you don't need it personally.

**Comment by doverride** [comment]
> > Copying contexts from foreign images is a bad idea and i am not going to allow portabled to do that. It was said: well this stuff already happens with network filesystems and networking. But there is one considerable difference: the network is not your local filesystem and a network filesystem is not your local filesystem either. When that unit is copied to the host, its the host policy that should have the last say.
> 
> No, it's not a bad idea, as it has been explained a million times already, so please stop with this absolutist statements. It doesn't fit your use case with untrusted images, which is perfectly fine (apart from the tiny detail that filesystems from untrusted images can exploit your kernel and all that, but that's life on Linux) and you are entitled to set up your images and your policies as you see fit. But you are not entitled to break our use case, which was not just validated but specifically requested by our SELinux experts, just because you don't need it personally.

In the end I did not break your use-case but you did break mine (3). We were too late I had (3) low on my wishlist  for a long time and if this patch was submitted say 6 months ago then it probably would have been a "fact of life" by now.

**Comment by poettering** [comment]
>  If you do 2. (like we do currently) then you have to tell the kernel (in detail what label to associate with the unit) and that only works if you have something to go by and in this case the unit is copied to a tmp file and then renamed. so you cannot just tell the kerne

Yeah, true. It's a general problem of the selinux approach of using the names of files as key for labelling them: it's systematically incompatible with the current approaches towards atomic file creation: creating files with temporary names and renaming them later...

**Comment by poettering** [comment]
it still appears to me that copying hte labels does make more sense. but as mentioned, happy to merge a patch that given some env var changes things to look up the labels in the db instead.

**Comment by doverride** [comment]
> > If you do 2. (like we do currently) then you have to tell the kernel (in detail what label to associate with the unit) and that only works if you have something to go by and in this case the unit is copied to a tmp file and then renamed. so you cannot just tell the kerne
> 
> Yeah, true. It's a general problem of the selinux approach of using the names of files as key for labelling them: it's systematically incompatible with the current approaches towards atomic file creation: creating files with temporary names and renaming them later...

Not too long ago we did not even have the flexibility of "name-based automatic transitions".Back then we only had file classes to use as key's. Nowaday's is quite a luxury to have both names as well as classes to use as key's and nine-out-of-ten that is enough.

I am just going to stick with what I currently have and block setfscreate access. I wont be able to give the portable units private contexts but they will be labeled as "generic units", so not suitable for confined administration but at least they're labeled "unit". Administration of portable units is limited so the main administrator and other confined services wont be able to control them either. In practice that is a reasonable compromise I  think.

**Comment by cgzones** [comment]
This pull request was designed as an alternative to #20767 and to illustrate a different approach. As #20767 was accepted I think this one has outlived its purpose. The main change would be to create `/etc/systemd/system.attached` with a label database lookup, which can also be realized via a name-based type transition in the policy. It seems to be better to let portabled be SELinux unaware for the time being (besides coping some *special* extended attribute).

So I'd like to withdraw this proposal.

**Comment by poettering** [comment]
OK, that's your prerogative.

(I still think making this controllable via env var would be ok, should you change your mind)

**Pull request closed without merging**